src/main/java/pl/jagiellonian/Models/PolynomialAsMap.java

```diff
@@ -37,16 +48,198 @@ public PolynomialAsMap multipleWith(PolynomialAsMap myPolynomial) {
 
                 List<Integer> monomialPowersResult = new ArrayList<>();
 
-                for (int i = 0; i < thisMonomialPowers.size(); i++) {
-                    monomialPowersResult.add(thisMonomialPowers.get(i) + monomialPowers.get(i));
+                int i = 0;
+                for (; i < thisMonomialPowers.size(); i++) {
+                    if (monomialPowers.size() > i) {
+                        monomialPowersResult.add(thisMonomialPowers.get(i) + monomialPowers.get(i));
+                    } else {
+                        monomialPowersResult.add(thisMonomialPowers.get(i));
+                    }
+                }
+                for (; i < monomialPowers.size(); i++) {
+                    monomialPowersResult.add(monomialPowers.get(i));
                 }
 
                 Integer monomialResultCoefficient = thisMonomialCoefficient * monomialCoefficient;
                 polynomialResult.put(monomialPowersResult, monomialResultCoefficient);
             }
         }
-        PolynomialAsMap polynomialAsMap = new PolynomialAsMap(polynomialResult);
-        return polynomialAsMap;
+        return new PolynomialAsMap(polynomialResult);
+    }
+
+    public PolynomialAsMap addPolynomial(PolynomialAsMap polynomialAsMap) {
+        Map<List<Integer>, Integer> polynomialToAddMap = new HashMap<>(polynomialAsMap.getPolynomialMap());
+        for (List<Integer> integers : polynomialMap.keySet()) {
+            if (polynomialToAddMap.get(integers) != null) {
+                polynomialToAddMap.put(integers, polynomialToAddMap.get(integers) + polynomialMap.get(integers));
+            } else {
+                polynomialToAddMap.put(integers, polynomialMap.get(integers));
+            }
+        }
+        return new PolynomialAsMap(polynomialToAddMap);
+    }
+
+    public void substitute(String expressionToReplace, String replacementExpressionString) {
+        if (!isSingleExpression(expressionToReplace)) {
+            throw new WrongFormatException();
+        }
+        ParsedExpression parsedExpressionToReplace = parseExpression(expressionToReplace);
+        List<String> variables = findVariables(replacementExpressionString);
+
+        ParsedExpression replacementExpression = parseExpression(replacementExpressionString);
+        PolynomialAsMap replacement = parsedExpressionToPolynomialAsMap(replacementExpression);
+        Set<Integer> indexesToReplace = parsedExpressionToReplace.getVariables();
+
+        List<PolynomialAsMap> fragments = new ArrayList<>();
+        VariablePowers:
+        for (List<Integer> variablePowers : polynomialMap.keySet()) {
+            int initialConstant = polynomialMap.get(variablePowers);
```

Comment:
substitute makes inefficient use of keySet iterator instead of entrySet iterator